maze_transformer/training/config.py

```diff
@@ -210,7 +216,81 @@ def transformer_config(self) -> HookedTransformerConfig:
             d_vocab=len(self.dataset_cfg.token_arr),
         )
 
+    def transformer_config(self) -> HookedTransformerConfig:
+        warnings.warn(
+            "transformer_config is deprecated, use hooked_transformer_cfg instead",
+            DeprecationWarning,
+        )
+        return self.hooked_transformer_cfg
+
     def create_model(self) -> HookedTransformer:
         return HookedTransformer(
-            cfg=self.transformer_config(), tokenizer=self.tokenizer
+            cfg=self.hooked_transformer_cfg,
+            tokenizer=self.tokenizer,
+        )
+
+    def create_model_zanj(self) -> ZanjHookedTransformer:
+        return ZanjHookedTransformer(self)
+
+
+@set_config_class(ConfigHolder)
+class ZanjHookedTransformer(ConfiguredModel, HookedTransformer):
+    """A hooked transformer that is configured by a ConfigHolder
+
+    the inheritance order is critical here -- super() does not call parent, but calls the next class in the MRO
+    So, we need ConfiguredModel to take the ConfigHolder and pass kwargs to HookedTransformer
+    """
+
+    def __init__(self, cfg_holder: ConfigHolder) -> None:
+        super().__init__(
+            # for ConfiguredModel
+            zanj_model_config=cfg_holder,
+            # for HookedTransformer
+            cfg=cfg_holder.hooked_transformer_cfg,
+            tokenizer=cfg_holder.tokenizer,
+        )
+
+    def _load_state_dict_wrapper(
+        self,
+        state_dict: dict[str, Any],
+        **kwargs,
+    ) -> None:
+        """this is a wrapper around the _load_state_dict function that allows us to do extra things when loading a state dict"""
+
+        recover_exact: bool = kwargs.get("recover_exact", False)
+        fold_ln: bool = kwargs.get("fold_ln", True)
```

Comment:
Why would we want to set either of these kwargs as part of model loading? They're both specified in the model config already right? What's the use case of changing these during the loading process?

Reply:
I am not entirely sure what the use case is, but this was a feature in the old `load_model_with_config`. It seemed unwise to remove the ability to fold the layernorm since I'm not sure what exactly it's for